File: pset2/crack/crack.c
//pset2/crack

#define _XOPEN_SOURCE
#include <unistd.h>
#include <stdio.h>
#include <string.h>

// TESTED HASHES:
// 25g4misjGGyG6 > 25MG
// 25LslphwWXHuc > 25zA
// anushree:50xcIMJ0y.RXo > YES
// brian:50mjprEcqC/ts > CA
// bjbrown:50GApilQSG3E2 > UPenn
// lloyd:50n0AAUD.pL8g > lloyd
// malan:50CcfIk1QrPr6 > maybe
// maria:509nVI8B9VfuA > TF
// natmelo:50JIIyhDORqMU > nope
// rob:50JGnXUgaafgc > ROFL
// stelios:51u8F0dkeDSbY > NO
// zamyla:50cI2vYkF0YU2 > LOL

int main(int argc, char **argv)
{
    // assuming user supplies a 56-bit DES hash
    if (argc == 2)
    {
        char password[6];
        char salt[3];
        strncpy(salt, argv[1], 2); // copy first two characters (bytes) of the hash into the salt
        // this is definitely a brute force / O(N5) solution
        // so this letter frequency might help add speed:
        // http://www.oxfordmathcenter.com/drupal7/node/353
        char *alpha = "\0eEtTaAoOiInNsShHrRdDlLcCuUmMwWfFgGyYpPbBvVkKjJsSqQzZ"; // set up a pointer
        // note: a previous attempt had an array of ASCII character numbers but this seemed to be impossibly slow - I wonder why?
        int len = 53; // there are 53 characters in alpha, including \0 :)
        // iterate through the entire max 5-digit array
        for (int i = 0; i < len; i++)
        {
            password[4] = alpha[i];
            for (int j = 0; j < len; j++)
            {
                password[3] = alpha[j];
                for (int k = 0; k < len; k++)
                {
                    password[2] = alpha[k];
                    for (int l = 0; l < len; l++)
                    {
                        password[1] = alpha[l];
                        for (int m = 0; m < len; m++)
                        {
                            // this is working backwards - so the most nested for loop deals with the [0] char of the password
                            // this means we test out one digit passwords before two digit, three digit etc.
                            password[0] = alpha[m];
                            // if the hashed password matches the user's inputted hash - bingo!
                            if (strcmp(argv[1], crypt(password, salt)) == 0)
                            {
                                printf("%s\n", password);
                                return 0;
                            }
                        }
                    }
                }
            }
        }
        // if the code reaches this stage it means a password hasn't been found
        printf("Uh oh! Password not found!\n");
        return 0;
    }
    else
    {
        printf("Usage: ./crack hash\n");
        return 1;
    }
}

File: pset6/crack/crack.py
#!/usr/bin/env python3
# pset6/crack.py

import crypt
import sys
from hmac import compare_digest as compare_hash

# TESTED HASHES:
# 25g4misjGGyG6 > 25MG
# 25LslphwWXHuc > 25zA
# anushree:50xcIMJ0y.RXo > YES
# brian:50mjprEcqC/ts > CA
# bjbrown:50GApilQSG3E2 > UPenn
# lloyd:50n0AAUD.pL8g > lloyd
# malan:50CcfIk1QrPr6 > maybe
# maria:509nVI8B9VfuA > TF
# natmelo:50JIIyhDORqMU > nope
# rob:50JGnXUgaafgc > ROFL
# stelios:51u8F0dkeDSbY > NO
# zamyla:50cI2vYkF0YU2 > LOL


def crack(needle):
    """Brute force O(n5) algorithm to crack up to a five character Unix DES-based hashed password"""
    if(needle):
        alpha = " eEtTaAoOiInNsShHrRdDlLcCuUmMwWfFgGyYpPbBvVkKjJsSqQzZ"
        alpha_length = len(alpha)
        needle = sys.argv[1]
        salt = needle[0:2]
        haystack = ["" for i in range(5)]
        for i in range(alpha_length):
            haystack[4] = alpha[i]
            for j in range(alpha_length):
                haystack[3] = alpha[j]
                for k in range(alpha_length):
                    haystack[2] = alpha[k]
                    for l in range(alpha_length):
                        haystack[1] = alpha[l]
                        for m in range(alpha_length):
                            haystack[0] = alpha[m]
                            # .strip() removes whitespace - very important here
                            current = "".join(haystack).strip()
                            if (compare_hash(crypt.crypt(current, salt), needle)):
                                return print(current)
        return("Uh oh! Password not found!")
    else:
        print("Usage: crack(password)")


if __name__ == "__main__":
    crack(sys.argv[1])
